### PR TITLE
Login endpoint

### PR DESCRIPTION
**Pull request by EdiOancea**

### CODE REVIEW

**Comment by EdiOancea** [comment]
[Login endpoint](https://trello.com/c/dXEfNHvW/4-login-endpoint)

**Comment by lgnstn** [comment]
@EdiOancea Please resolve conflicts.

**Comment by EdiOancea** [comment]
@lgnstn updated

File: app/services/signin.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+const signinRepository = require('../repositories/signin');
+const errorService = require('../services/error');
+const privateKey = process.env.API_KEY;
+
+class SigninService {
+  async authenticate({ email, password }) {
+    const user = await signinRepository.getUserPassword(email);
+    if (!user) {
+      errorService.throwValidationError("Invalid email.");
```
Comment (lgnstn) [inline app/services/signin.js]: You don't want to give details about why the authentication didn't succeeded (security reasons). All the errors should say "Wrong credentials" or something like this.

File: app/services/signin.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    }
+
+    const res = await bcrypt.compare(password, user.password);
+    if (!res) {
+      errorService.throwValidationError("Invalid password.");
+    }
+    
+    return await jwt.sign({ id: user.id }, privateKey, { expiresIn: 3600 });
```
Comment (lgnstn) [inline app/services/signin.js]: I think this is a sync operation, you don't need `await`.

File: app/controllers/signin.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+const signinService = require('../services/signin');
+
+class SigninController {
+  async authenticate(req, res, next) {
+    res.json(await signinService.authenticate(req.body));
```
Comment (lgnstn) [inline app/controllers/signin.js]: Please use an envelope like this `{ token: await signinService.authenticate(req.body) }`.
 
Also good to know: https://blog.mwaysolutions.com/2014/06/05/10-best-practices-for-better-restful-api/

File: test/signin.test.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
+const request = require('supertest');
+const expect = require('chai').expect;
+const jwt = require('jsonwebtoken');
+
+const app = require('../index');
+const userFixture = require('../fixtures/user');
+const loadFixture = require('./loadFixture');
+const privateKey = process.env.API_KEY;
```
Comment (lgnstn) [inline test/signin.test.js]: Please rename it `JWT_SECRET` and add a `.env.sample` file with dummy data for all the env variables (basically a copy of `.env` with dummy data for sensitive info).

**Review by lgnstn — CHANGES_REQUESTED** [review]

**Comment by EdiOancea** [comment]
@lgnstn updated

**Review by lgnstn — APPROVED** [review]

**Pull request merged**